### ISSUE TITLE
The accuracy loss in the CUDA version is much more than Triton version (Llama-3.2-1B)

### ISSUE THREAD

**Issue by WanliZhong** [thu-ml/SageAttention#154]
> Env: Ubuntu 22.04, python 3.11, CUDA 12.8, pytorch preview, pytorch-triton  3.3.0+git96316ce5
> NVIDIA RTX 5880 Ada Generation, sm89
> compile log: 

### Test results

Baseline without sageattention
```text
piqa
alias: piqa
acc,none: 0.7453754080522307
acc_stderr,none: 0.01016443223706048
acc_norm,none: 0.7464635473340587
acc_norm_stderr,none: 0.01015009083455177

wikitext
alias: wikitext
word_perplexity,none: 11.568933844064432
word_perplexity_stderr,none: N/A
byte_perplexity,none: 1.5806698488959285
byte_perplexity_stderr,none: N/A
bits_per_byte,none: 0.6605360665579088
bits_per_byte_stderr,none: N/A
```

sageattn_qk_int8_pv_fp16_triton
```text
piqa
alias: piqa
acc,none: 0.7415669205658324
acc_stderr,none: 0.010213971636773326
acc_norm,none: 0.7464635473340587
acc_norm_stderr,none: 0.01015009083455177

wikitext
alias: wikitext
word_perplexity,none: 11.587716504315202
word_perplexity_stderr,none: N/A
byte_perplexity,none: 1.5811494406605209
byte_perplexity_stderr,none: N/A
bits_per_byte,none: 0.6609737289303412
bits_per_byte_stderr,none: N/A
```

**sageattn_qk_int8_pv_fp16_cuda**:  The results significantly decreased or were incorrect.

```text
piqa
alias: piqa
acc,none: 0.5810663764961915
acc_stderr,none: 0.011511473119995072
acc_norm,none: 0.5854189336235038
acc_norm_stderr,none: 0.011494326682255163

wikitext
alias: wikitext
word_perplexity,none: nan
word_perplexity_stderr,none: N/A
byte_perplexity,none: nan
byte_perplexity_stderr,none: N/A
bits_per_byte,none: nan
bits_per_byte_stderr,none: N/A
```

### Reproduce code
```python
import torch
from lm_eval import evaluator  # EleutherAI lm-evaluation-harness
from lm_eval.models.huggingface import HFLM 

import torch.nn.functional as F
import sageattention
from functools import partial

F.scaled_dot_product_attention = partial(
     sageattention.sageattn_qk_int8_pv_fp16_cuda,
#     sageattention.sageattn_qk_int8_pv_fp16_triton,
     pv_accum_dtype="fp32")

# F.scaled_dot_product_attention = sageattention.sageattn

def evaluate_model(model_name, batch_size=1):
    tasks = ["piqa", "wikitext"]

    hf_model = HFLM(
        pretrained=model_name,
        batch_size=batch_size,
        dtype=torch.float16,
        device_map="auto", 
        # attn_implementation="flash_attention_2",
                    )

    eval_results = evaluator.simple_evaluate(
        model=hf_model,
        tasks=tasks,
    )
    return eval_results

def main():
    # MODEL_NAME = "meta-llama/Llama-3.1-8B"
    MODEL_NAME = "meta-llama/Llama-3.2-1B"

    print("Evaluating model...")
    eval_results = evaluate_model(MODEL_NAME)

    for task in eval_results["results"]:
        print(f"\n{task}")
        for item in eval_results["results"][task]:
            print(f"{item}: {eval_results['results'][task][item]}")

if __name__ == "__main__":
    main()

```

**Comment by jason-huang03** [comment]
@WanliZhong We believe this is due to the evaluating process of lm_eval and currently we cannot figure out what happens. Using our own script to evaluate wikitext perplexity, the result is normal. We provide the script as follows:

```python

import os
import argparse, tqdm
from datasets import load_dataset
import json
import torch
import torch.nn as nn
from transformers import AutoTokenizer, AutoConfig, AutoModelForCausalLM

LENGTH = 2048  # parameter for ppl 
class Evaluator:
    def __init__(self, dataset, tokenizer, device, n_samples=40):
        self.dataset = dataset
        self.tokenizer = tokenizer
        self.device = device

        self.dataset = tokenizer(
            "\n\n".join(dataset["text"]), return_tensors="pt"
        ).input_ids.to(device)

        self.n_samples = n_samples

    @torch.no_grad()
    def evaluate(self, model):
        model.eval()
        nlls = []
        for i in tqdm.tqdm(range(self.n_samples), desc="Evaluating..."):
            batch = self.dataset[:, (i * LENGTH) : ((i + 1) * LENGTH)].to(model.device)
            with torch.no_grad():
                lm_logits = model(batch).logits
            shift_logits = lm_logits[:, :-1, :].contiguous().float()
            shift_labels = self.dataset[:, (i * LENGTH) : ((i + 1) * LENGTH)][:, 1:]
            loss_fct = nn.CrossEntropyLoss()
            loss = loss_fct(
                shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1)
            )
            neg_log_likelihood = loss.float() * LENGTH
            nlls.append(neg_log_likelihood)

        return torch.exp(torch.stack(nlls).sum() / (self.n_samples * LENGTH))
    

def parse_args():
    parser = argparse.ArgumentParser(description="LLM Evaluation")
    parser.add_argument(
        "--model_name", required=True, type=str, help="Path of the model"
    )
    parser.add_argument(
        "--dtype", type=str, choices=["fp16", "fp32", "bf16"], default="fp16"
    )
    # tokenize related
    parser.add_argument(
        "--max_length", default=2048, type=int, help="Maximum token length"
    )
    # flash attention
    parser.add_argument(
        "--use_flash_attention",
        action="store_true",
        default=False,
        help="Whether to use flash attention",
    )

    args = parser.parse_args()
    return args

if __name__ == "__main__":
    truncation = True
    last_words = False

    args = parse_args()

    print(args)

    # load tokenizer
    tokenizer = AutoTokenizer.from_pretrained(
        args.model_name, trust_remote_code=True, use_fast=True
    )
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token
    if tokenizer.pad_token_id is None:
        tokenizer.pad_token_id = tokenizer.eos_token_id

    # define model config
    config = AutoConfig.from_pretrained(args.model_name, trust_remote_code=True)
    if args.use_flash_attention:
        config._attn_implementation = "sdpa"
    else:
        config._attn_implementation_internal = "eager"

    print(f"using {config._attn_implementation_internal} attention implementation")

    if args.dtype == 'fp16':
        dtype = torch.float16
    elif args.dtype == 'fp32':
        dtype = torch.float32
    elif args.dtype == 'bf16':
        dtype = torch.bfloat16
    else:
        raise ValueError(f"unsupported type: {args.dtype}")

    # load model
    model = AutoModelForCausalLM.from_pretrained(
        args.model_name,
        config=config,
        device_map="auto",
        torch_dtype=dtype,
        trust_remote_code=True,
        attn_implementation=(
            "eager" if not args.use_flash_attention else "sdpa"
        ),
    ).eval()

    F.scaled_dot_product_attention = partial(
        sageattention.sageattn_qk_int8_pv_fp16_cuda,
    #     sageattention.sageattn_qk_int8_pv_fp16_triton,
        pv_accum_dtype="fp32")
    
    # Evaluate Wikitext ppl
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    dataset = load_dataset('Salesforce/wikitext', 'wikitext-2-raw-v1', split='test')
    evaluator = Evaluator(dataset, tokenizer, device)
    ppl_fp16 = evaluator.evaluate(model)
    print(f"Perplexity: {ppl_fp16}")
```

running `python test.py --model_name llama_model_path--use_flash_attention --dtype bf16` and there will be the result.

**Comment by WanliZhong** [comment]
@jason-huang03 Thanks! I will try your script to run!
